### PR TITLE
Do not return duplicate data from `component_data_objects` or `component_data_iterindex`

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This is the first step toward resolving issues with references where the model component iterators can return the same data multiple times.  The implementation is aware that the only way to see the same component twice is through `Reference`s, and makes use of this to minimize the overhead of deduplication (in particular, `id()`s are only cached for component containers, _unless_ a `Reference` is encountered).

This implementation has only minimal impact on performance (<1% overhead):
image

Note that this does NOT resolve the issues in the writers, etc.  That will be implemented in a subsequent PR

## Changes proposed in this PR:
- `component_data_objects()` and `component_data_iterindex()` will not yield the same data object more than once
- clean up GDP handling of deactivated blocks / disjuncts 
- updated documentation

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/block.py [hunk 1]
```diff
@@ -1334,10 +1414,15 @@ def _component_typemap(self, ctype=None, active=None, sort=False):
         else:
             return PseudoMap(self, ctype, active, sort)
 
-    def _component_data_iter(self, ctype=None, active=None, sort=False):
-        """
-        Generator that returns a 3-tuple of (component name, index value,
-        and _ComponentData) for every component data in the block.
+    def _component_data_iteritems(self, ctype, active, sort, dedup):
```
Comment (michaelbynum) [inline pyomo/core/base/block.py:1417]: Can you please add the expected types of the arguments (either as type hints or in the doc string)? I believe deduce is expected to be an instance of `_DeduplicateInfo`, but future me (if I ever look at this code again) will not remember that.

**Review by michaelbynum — COMMENTED** [review]
I am super excited about this PR. I also really like this implementation. It should add very little overhead. I do have a few minor comments/questions.

**Comment by michaelbynum** [comment]
P.s. Thanks for checking the performance.

File: pyomo/core/base/block.py [hunk 2]
```diff
@@ -120,6 +120,84 @@ def __len__(self):
...
+                = self.seen_comp_thru_reference.__contains__
+            seen_comp_thru_reference_add = self.seen_comp_thru_reference.add
+            seen_data_contains = self.seen_data.__contains__
+            seen_data_add = self.seen_data.add
+
+            for _item in items:
+                _data = _item if are_values else _item[1]
+                # If the data is contained in a component we have
+                # already processed, then it is a duplicate and we can
+                # bypass forther checks.
```
Comment (blnicho) [inline pyomo/core/base/block.py:163]: Typo: further

File: pyomo/core/base/block.py [hunk 3]
```diff
@@ -1482,72 +1592,56 @@ def block_data_objects(self,
                            sort=False,
                            descend_into=True,
                            descent_order=None):
+        """Generator returning this block and an any matching sub-blocks.
```
Comment (blnicho) [inline pyomo/core/base/block.py:1595]: Typo: "and an any"

File: pyomo/core/base/block.py [hunk 4]
```diff
@@ -1482,72 +1592,56 @@ def block_data_objects(self,
...
+        -----
+        The `self` block is *always* returned, regardless of the types
+        indicated by descend_into.
 
-        # TODO: merge into block_data_objects
-        if ctype is None:
-            ctype = (Block,)
-        elif isclass(ctype):
-            ctype = (ctype,)
+        The active flag is enforced on *all* blocks, inclufing self.
```
Comment (blnicho) [inline pyomo/core/base/block.py:1609]: Typo: including

**Review by emma58 — COMMENTED** [review]
This is really cool! I don't have any questions Michael didn't already ask...

File: pyomo/core/base/block.py [hunk 5]
```diff
@@ -1482,72 +1592,56 @@ def block_data_objects(self,
...
 
-        """
-        This method returns a generator that iterates
-        through the current block and recursively all
-        sub-blocks.  This is semantically equivalent to
+        This is roughly equivalent to
 
-            component_data_objects(Block, ...)
+            iter(block for block in itertools.chain(
+                [self], self.component_data_objects(descend_into, ...))
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/core/base/block.py:1600]: Not really - `descend_into` provides the `ctype`; if it is `True`, it is mapped to `(Block,)`.  This is also why I introduced it as "roughly equivalent to".  To make it exact for all cases, I would end up replicating the entire body of the function.  I

File: pyomo/core/base/indexed_component.py [hunk 6]
```diff
@@ -403,8 +403,7 @@ def keys(self, ordered=False):
 
         """
         sort_needed = ordered
-        if hasattr(self._index_set, 'isfinite') and not \
-           self._index_set.isfinite():
+        if not self._index_set.isfinite():
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/core/base/indexed_component.py:406]: I think this was legacy code from before the Set rewrite.  I could not see any path / test / case where the `hasattr` test was needed.

**Review by blnicho — APPROVED** [review]

**Review by michaelbynum — APPROVED** [review]

**Pull request merged**